### ISSUE TITLE
Please update training layers, add an option to train on dense layers

### ISSUE THREAD

**Issue by RandomInternetPreson** [oobabooga/text-generation-webui#3637]
**Description**

This is a feature request that is echoed here: 

https://github.com/huggingface/peft/issues/735

A way to specify dense layers when training please.

I can edit line 49 of training.py from:

 standard_modules = ["q_proj", "v_proj"]

to this:

 standard_modules = ["q_proj", "v_proj","o_proj","gate_proj","up_proj","k_proj","down_proj"]

as a workaround but it took me a while to figure out what to do and I was editing the wrong code (other.py in perf/util)

**Additional Context**

For max. performance, the QLoRA paper suggests that LoRA should be applied to all linear layers:

As shown in Figure 2 for LLaMA 7B finetuning on Alpaca, we find that the most critical LoRA hyper-parameter is how many LoRA adapters are used in total and that LoRA on all linear transformer block layers are required to match full finetuning performance.

Doing all the projections should yield a better quality LoRA

Check out the paper here on page 6:
https://arxiv.org/pdf/2305.14314.pdf

The RougeL metric is near identical to the fp16 model:
"Using LoRA on all transformer layers is critical to
match 16-bit performance."

**Comment by hypersniper05** [comment]
Thank you for this. I will make the changes manually too. I think training has been getting a few upgrades on here but if they focus a little more on the training side they would see an increase of users using the platform especially since the ui is getting easier and easier to use

**Comment by hypersniper05** [comment]
> 

Can you share which training.py? I do see those parameters in it

**Comment by RandomInternetPreson** [comment]
Sure, it's the training.py file in the modules folder:
https://github.com/oobabooga/text-generation-webui/tree/main/modules

**Comment by RandomInternetPreson** [comment]
Welp, it looks like the code was updated not too long ago.  Now I don't know if the newest version does, I'll have to install and test it out.  Perhaps the desired enhancement is already implemented, I'm running a version of oobabooga from aug 9.

**Comment by hypersniper05** [comment]
The new version only loads the q_ v_ , it tells you in the command window when starting training :(

**Comment by hypersniper05** [comment]
[https://github.com/oobabooga/text-generation-webui/commit/b96fd22a81df547c32819ce846d8517cde1153f2]()

**Comment by RandomInternetPreson** [comment]
Thanks for the confirmation, I was predicting that based off this:

https://github.com/huggingface/peft/blob/main/src/peft/utils/other.py

When you install oobabooga this is one of the packages that gets installed, this "other.py" file located here:
installer_files/env/lib/site-packages/peft/utils

This is the .py file that the training.py file is referencing, can you try two things for me?

1.  Go to line 385 of this other.py file it should look like this:
"llama": ["q_proj", "v_proj"],
and change it to this
"llama": ["q_proj", "v_proj","o_proj","gate_proj","up_proj","k_proj","down_proj"],

2.  If that doesn't work try going to your training.py file and look for this line (maybe line 27)

from peft.utils.other import \
    TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING as model_to_lora_modules
from transformers.models.auto.modeling_auto import (
    MODEL_FOR_CAUSAL_LM_MAPPING_NAMES
)

and replace this with:
try:
    from peft.utils.other import \
        TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING as \
        model_to_lora_modules
    from transformers.models.auto.modeling_auto import (
        MODEL_FOR_CAUSAL_LM_MAPPING_NAMES
    )
    MODEL_CLASSES = {v: k for k, v in MODEL_FOR_CAUSAL_LM_MAPPING_NAMES}
except:
    standard_modules = ["q_proj", "v_proj","o_proj","gate_proj","up_proj","k_proj","down_proj"]
    model_to_lora_modules = {"llama": standard_modules, "opt": standard_modules, "gptj": standard_modules, "gpt_neox": ["query_key_value"], "rw": ["query_key_value"]}
    MODEL_CLASSES = {
        "LlamaForCausalLM": "llama",
        "OPTForCausalLM": "opt",
        "GPTJForCausalLM": "gptj",
        "GPTNeoXForCausalLM": "gpt_neox",
        "RWForCausalLM": "rw"

    }

**Comment by RandomInternetPreson** [comment]
[training.txt](https://github.com/oobabooga/text-generation-webui/files/12401706/training.txt)


I've updated the training.py file for you just in case the formatting I provided doesn't come across well. 

Copy this training.txt file and rename it to training.py and replace the training.py file in the modules folder (back up the original training.py file first though).

**Comment by RandomInternetPreson** [comment]
If that training.txt file doesn't do the trick try this one: 
[training2.txt](https://github.com/oobabooga/text-generation-webui/files/12401750/training2.txt)

**Comment by RandomInternetPreson** [comment]
[training3.txt](https://github.com/oobabooga/text-generation-webui/files/12401763/training3.txt)

and maybe third time is the charm, I'll need to look at things more closely if this one doesn't work, but give it a try if the previous fixes don't help, and let me know what did or did not work for you.

**Comment by hypersniper05** [comment]
> 1. "q_proj", "v_proj","o_proj","gate_proj","up_proj","k_proj","down_proj"
It Worked
<img width="862" alt="image" src="https://github.com/oobabooga/text-generation-webui/assets/29990615/09595dce-0770-4917-a350-e72485c03b45">


looks like other.py did the trick. Its now line 220 that needs to be changed:

<img width="663" alt="image" src="https://github.com/oobabooga/text-generation-webui/assets/29990615/562bc733-a800-4ed1-a05f-f2691b195b9d">

**Comment by hypersniper05** [comment]
yeah for sure! it was at 7% only now its at 27% . I had to lower my micro batch size but its ok . Thanks for the help

**Comment by Ph0rk0z** [comment]
Yea, this should be an option so you can do a simple lora or a more full finetune.

**Comment by hypersniper05** [comment]
I just came back to say that my fine-tuned model is way more responsive in the way I trained it than before making this change. This is a must!

**Comment by RandomInternetPreson** [comment]
Yeass!  Very interesting, thank you for the extra info, I've noticed this too <3 

I really hope they add an option to do the dense training, I think people need to have the opportunity to test out different training schemes.  Increasing the training density uses more vram, but it might use it in a more beneficial way depending on training needs.

**Comment by lighteternal** [comment]
This is a very important option for finetuning. 👍

**Comment by oobabooga** [comment]
v4.0 adds a "Target all linear layers" checkbox that applies LoRA to every `nn.Linear` layer (except `lm_head`), which is what you were asking for. See the [release notes](https://github.com/oobabooga/text-generation-webui/releases/tag/v4.0) and [training docs](https://github.com/oobabooga/text-generation-webui/wiki/05-%E2%80%90-Training-Tab).
